Sources/Fuzzilli/Core/ProgramBuilder.swift

```diff
@@ -335,27 +330,39 @@ public class ProgramBuilder {
     }
     
     /// Maps a variable from the program that is currently configured for adoption into the program being constructed.
-    public func adopt(_ variable: Variable, keepType: Bool) -> Variable {
+    public func adopt(_ variable: Variable) -> Variable {
         if !varMaps.last!.contains(variable) {
             varMaps[varMaps.count - 1][variable] = nextVariable()
         }
-        let currentVariable = varMaps.last![variable]!
 
-        if keepType, let currentType = typeMaps.last![variable] {
-            program.setRuntimeType(of: currentVariable, to: currentType)
-        }
-        return currentVariable
+        return varMaps.last![variable]!
     }
     
     /// Maps a list of variables from the program that is currently configured for adoption into the program being constructed.
-    public func adopt(_ variables: [Variable], keepTypes: Bool) -> [Variable] {
-        return variables.map{ adopt($0, keepType: keepTypes) }
+    public func adopt(_ variables: [Variable]) -> [Variable] {
+        return variables.map(adopt)
+    }
+
+    private func adoptTypes(_ originalOutputs: ArraySlice<Variable>, _ adoptedOutputs: ArraySlice<Variable>) {
+        for (originalVariable, adoptedVariable) in zip(originalOutputs, adoptedOutputs) {
+            if let type = typeMaps.last![originalVariable] {
+                program.setRuntimeType(of: adoptedVariable, to: type)
+
+                if type != .unknown {
+                    interpreter?.setType(of: adoptedVariable, to: type)
+                }
+            }
+        }
     }
     
     /// Adopts an instruction from the program that is currently configured for adoption into the program being constructed.
     public func adopt(_ instruction: Instruction, keepTypes: Bool) {
-        let newInouts = adopt(Array(instruction.inputs), keepTypes: false) + adopt(Array(instruction.allOutputs), keepTypes: keepTypes)
-        internalAppend(Instruction(operation: instruction.operation, inouts: newInouts))
+        let newInouts = adopt(Array(instruction.inputs)) + adopt(Array(instruction.allOutputs))
+        let adoptedInstruction = Instruction(operation: instruction.operation, inouts: newInouts)
+        internalAppend(adoptedInstruction)
+        if keepTypes {
+            adoptTypes(instruction.allOutputs, adoptedInstruction.allOutputs)
```

Comment:
Would it be a bit nicer if the function signature was `adoptTypes(from origInstr: Instruction, to newInstr: Instruction)`?